WordPress/gutenberg
# BaseControl: Migrate to TypeScript

Part of #35744 

## What?

Converts the `BaseControl` component to TypeScript.

## Why?

Improves devex and code quality through type safety, IDE type hints, and better Storybook documentation.

## How?

- The readme and type data are now in sync. Some parts were edited for clarity.
- No runtime changes.

## Testing Instructions

`npm run storybook:dev` and check the Docs tab for `BaseControl`. It should be useful and understandable.

For easier code review, [exclude the commit](https://github.com/WordPress/gutenberg/pull/39468/files/1da384e3b4d338506179ae99de5f5d6867e99a50..f1eade4b67b2e343b5cc0fb4269e72c425d1eccb) where I just changed the file extensions.

✅ Tests pass
